### PR TITLE
Ipv6 ipvs proto tcp/udp/icmp

### PR DESCRIPTION
**Pull request by xiangp126**

### CODE REVIEW

File: include/ipv6.h [hunk 1]
```diff
@@ -32,6 +32,14 @@
 
 #define IPV6
 #define RTE_LOGTYPE_IPV6    RTE_LOGTYPE_USER1
+#define IPV6_HDR_FIXED_LEN  40
```
Comment (mscbg) [inline include/ipv6.h]: delete useless code

File: src/ipvs/ip_vs_proto_udp.c [hunk 2]
```diff
@@ -441,19 +545,26 @@ static int udp_fnat_in_handler(struct dp_vs_proto *proto,
                     struct rte_mbuf *mbuf)
 {
     struct udp_hdr *uh = NULL;
-    struct iphdr *iph = (void *)ip4_hdr(mbuf);
+    struct iphdr  *iph = NULL;
```
Comment (AdamYYan) [inline src/ipvs/ip_vs_proto_udp.c]: 直接定义为void * 类型是否更恰当?

File: src/ipvs/ip_vs_proto_udp.c [hunk 3]
```diff
@@ -348,28 +411,51 @@ static int insert_opp_uoa(struct dp_vs_conn *conn, struct rte_mbuf *mbuf,
...
+        opph->version  = 0x2;
+        opph->protocol = ((struct ip6_hdr *)niph)->ip6_nxt;
+    } else {
+        /* version 1 for ipv4 address */
+        opph->version  = 0x1;
+        opph->protocol = ((struct iphdr *)niph)->protocol;
+    }
+    opph->length = htons(sizeof(*opph) + IPOLEN_UOA);
 
+    memset(uoa, 0, sizeof(struct ipopt_uoa));
```
Comment (AdamYYan) [inline src/ipvs/ip_vs_proto_udp.c]: uoa指向null？

File: src/ipvs/ip_vs_proto_udp.c [hunk 4]
```diff
@@ -190,37 +193,66 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
+        ((struct iphdr *)iph)->daddr   = conn->daddr.in.s_addr;
+    }
 
     if (mode == UOA_M_IPO) {
-        iph->ihl    = (sizeof(struct iphdr) + IPOLEN_UOA) / 4;
-        iph->tot_len = htons(sizeof(*iph) + IPOLEN_UOA + sizeof(*uh));
-        iph->protocol = oiph->protocol; /* should always UDP */
+        /* only ipv4 support and use ip option mode, and thus get here */
+        ((struct iphdr *)iph)->ihl =
+            (sizeof(struct iphdr) + IPOLEN_UOA) / 4;
```
Comment (AdamYYan) [inline src/ipvs/ip_vs_proto_udp.c]: IPOLEN-uoa是否合适？

File: src/ipvs/ip_vs_proto_udp.c [hunk 5]
```diff
@@ -190,37 +193,66 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
+        ((struct iphdr *)iph)->ttl     = ((struct iphdr *)oiph)->ttl;
+        ((struct iphdr *)iph)->saddr   = conn->laddr.in.s_addr;
+        ((struct iphdr *)iph)->daddr   = conn->daddr.in.s_addr;
+    }
 
     if (mode == UOA_M_IPO) {
-        iph->ihl    = (sizeof(struct iphdr) + IPOLEN_UOA) / 4;
-        iph->tot_len = htons(sizeof(*iph) + IPOLEN_UOA + sizeof(*uh));
-        iph->protocol = oiph->protocol; /* should always UDP */
+        /* only ipv4 support and use ip option mode, and thus get here */
```
Comment (AdamYYan) [inline src/ipvs/ip_vs_proto_udp.c]: 是否需要代码保证下只支持ipv4

File: src/ipvs/ip_vs_proto_icmp.c [hunk 6]
```diff
@@ -72,11 +74,15 @@ static int icmp_conn_sched(struct dp_vs_proto *proto,
...
+    /* cast to struct icmphdr * or struct icmp6_hdr * before used  */
+    void *ich = NULL, *_icmph = NULL;
     struct dp_vs_service *svc;
+    int af = iph->af;
+    int icmphdr_len = (af == AF_INET6) ?
+                    sizeof(struct icmp6_hdr) : sizeof(struct icmphdr);
     assert(proto && iph && mbuf && conn && verdict);
 
-    ich = mbuf_header_pointer(mbuf, iph->len, sizeof(_icmph), &_icmph);
+    ich = mbuf_header_pointer(mbuf, iph->len, icmphdr_len, _icmph);
```
Comment (mscbg) [inline src/ipvs/ip_vs_proto_icmp.c]: _icmph is not malloc. will coredump in mbuf_header_pointer.

File: src/ipvs/ip_vs_proto_udp.c [hunk 7]
```diff
@@ -230,10 +262,16 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
-    uoa->op_len     = IPOLEN_UOA;
-    uoa->op_port    = ouh->source;
-    uoa->op_addr    = oiph->saddr;
+    uoa->op_code = IPOPT_UOA;
+    uoa->op_len  = IPOLEN_UOA;
+    uoa->op_port = ouh->source;
+    /* fix uoa->op_addr */
+    if (af == AF_INET6) {
+        memcpy(uoa->op_addr, &((struct ip6_hdr *)oiph)->ip6_src, 
+                                                    IPV6_ADDR_LEN_IN_BYTES);
```
Comment (AdamYYan) [inline src/ipvs/ip_vs_proto_udp.c]: 对齐格式

File: src/ipvs/ip_vs_proto_icmp.c [hunk 8]
```diff
@@ -102,58 +108,117 @@ static int icmp_conn_sched(struct dp_vs_proto *proto,
...
+                || type == ICMP_ADDRESSREPLY)
+      return true;
+    else
+      return false;
+}
+
+static bool is_icmp6_reply(uint8_t type) {
+    if (type == ICMP6_ECHO_REPLY) {
+        return true;
+    }
```
Comment (mscbg) [inline src/ipvs/ip_vs_proto_icmp.c]: other type is not support yet? SNAT need other type

File: src/ipvs/ip_vs_proto_icmp.c [hunk 9]
```diff
@@ -102,58 +108,117 @@ static int icmp_conn_sched(struct dp_vs_proto *proto,
...
+    int af = iph->af;
+    /* true icmp type/code, used for v4/v6 */
+    uint8_t icmp_type = 0;
+    uint8_t icmp_code = 0;
+    int icmphdr_len = (af == AF_INET6) ?
+                    sizeof(struct icmp6_hdr) : sizeof(struct icmphdr);
     assert(proto && iph && mbuf);
 
-    ich = mbuf_header_pointer(mbuf, iph->len, sizeof(_icmph), &_icmph);
+    ich = mbuf_header_pointer(mbuf, iph->len, icmphdr_len, _icmph);
```
Comment (mscbg) [inline src/ipvs/ip_vs_proto_icmp.c]: coredump

File: src/ipvs/ip_vs_proto_udp.c (hunk 3 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: 这个memset 是可以的，因为地址空间在前面原334行已经申请好了
```c
 334     niph = (void *)rte_pktmbuf_prepend(mbuf, sizeof(*opph) + IPOLEN_UOA);
```
对uoa的memset同下面的opph
```c
358      opph = (struct opphdr *)((void *)niph + (niph->ihl << 2));
359      memset(opph, 0, sizeof(*opph));
```

File: include/uoa.h [hunk 10]
```diff
@@ -35,13 +35,20 @@
...
+ * @op_len: option length
+ * @op_port: port number
+ * @op_addr: ipv4/v6 address
+ */
 struct ipopt_uoa {
 	__u8	op_code;
 	__u8	op_len;
 	__be16  op_port;
-	__be32  op_addr;
+	__be32  op_addr[4];
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply include/uoa.h]: IPOLEN_UOA 就是定义成结构体长度
```c
#define IPOLEN_UOA  sizeof(struct ipopt_uoa)
```

File: src/ipvs/ip_vs_proto_udp.c (hunk 7 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: 这个怎么对齐好像也不是很美观，目前这样也没有超过每行80字节
可以拉到前面和第二参数竖对齐

File: include/ipvs/proto.h [hunk 11]
```diff
@@ -24,6 +24,8 @@
 #include "ipvs/conn.h"
 
 struct dp_vs_conn;
+#define IPV6_ADDR_LEN_IN_BYTES 16
+#define IPV4_ADDR_LEN_IN_BYTES 4
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply include/ipvs/proto.h]: ok

File: src/ipvs/ip_vs_proto_tcp.c [hunk 12]
```diff
@@ -95,44 +95,62 @@ static struct tcp_state tcp_states[] = {
...
 inline void tcp4_send_csum(struct ipv4_hdr *iph, struct tcphdr *th)
 {
     th->check = 0;
     th->check = rte_ipv4_udptcp_cksum(iph, th);
 }
 
-inline void tcp6_send_csum(struct ipv6_hdr *ip6h, struct tcphdr *th)
-{
+/*
+ * tcp6_send_csum: compute checksum for tcp/udp ipv6
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_tcp.c]: ok，我把udp去掉

File: src/ipvs/ip_vs_proto_icmp.c (hunk 6 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_icmp.c]: ok，是有这个问题

[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_icmp.c]: 根据王庆提出的问题这里已改

File: src/ipvs/ip_vs_proto_icmp.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_icmp.c]: 同上

File: src/ipvs/ip_vs_proto_tcp.c [hunk 13]
```diff
@@ -573,16 +591,17 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
     struct tcphdr *th;
     struct route_entry *rt = mbuf->userdata;
     struct netif_port *dev = NULL;
-    int ip4hlen = ip4_hdrlen(mbuf);
-    
-    if (mbuf_may_pull(mbuf, ip4hlen + sizeof(*th)) != 0)
+    int af = conn->af;
+    int iphdrlen = ((af == AF_INET6) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_tcp.c]: 修改之后的函数包含获取拓展包头长度，是包装了一下

File: src/ipvs/ip_vs_proto_tcp.c [hunk 14]
```diff
@@ -795,12 +870,14 @@ static int tcp_state_trans(struct dp_vs_proto *proto, struct dp_vs_conn *conn,
...
     unsigned conn_timeout = 0;
+    int af = conn->af;
 #ifdef CONFIG_DPVS_IPVS_DEBUG
     char dbuf[64], cbuf[64];
     const char *daddr, *caddr;
 #endif
 
-    th = mbuf_header_pointer(mbuf, ip4_hdrlen(mbuf), sizeof(_tcph), &_tcph);
+    int iphdrlen = ((af == AF_INET6) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
+    th = mbuf_header_pointer(mbuf, iphdrlen, sizeof(_tcph), &_tcph);
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_tcp.c]: 重新修改了ip6_hdrlen，保证接口一致
```c
809 /*
810  * ip6_hdrlen: get ip6 header length, including extension header length
811  */
812 int ip6_hdrlen(const struct rte_mbuf *mbuf) {
813     struct ip6_hdr *ip6h = ip6_hdr(mbuf);
814     uint8_t ip6nxt = ip6h->ip6_nxt;
815     int ip6_hdrlen = ip6_skip_exthdr(mbuf, mbuf->l3_len, &ip6nxt);
816     if (ip6_hdrlen < 0) {
817         return EDPVS_INVPKT;
818     }
819     return ip6_hdrlen;
820 }
```

File: src/ipvs/ip_vs_proto_udp.c [hunk 15]
```diff
@@ -168,16 +170,17 @@ static int udp_state_trans(struct dp_vs_proto *proto, struct dp_vs_conn *conn,
 
 static int send_standalone_uoa(const struct dp_vs_conn *conn,
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: 147行，已修改
```c
140     /*
141      * L2 confirm neighbour
142      * UDP has no ack, we don't know pkt from client is response or not
143      * UDP can only confirm neighbour to RS
144      */
145     int af = iph->af;
146     if (conn != NULL) {
147         if (AF_INET6 == af) {
148             if ((*direct == DPVS_CONN_DIR_OUTBOUND) && conn->in_dev
149                         && !ipv6_addr_any(&conn->in_nexthop.in6)) {
150                 neigh_confirm(AF_INET6, &conn->in_nexthop, conn->in_dev);
151             }
152         } else {
153             if ((*direct == DPVS_CONN_DIR_OUTBOUND) && conn->in_dev
154                         && (conn->in_nexthop.in.s_addr != htonl(INADDR_ANY))){
155                 neigh_confirm(AF_INET, &conn->in_nexthop, conn->in_dev);
156             }
157         }
158     }
```

File: src/ipvs/ip_vs_proto_udp.c (hunk 2 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: ok,本意就是要定义成void *

File: src/ipvs/ip_vs_proto_tcp.c [hunk 16]
```diff
@@ -617,14 +636,28 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
...
+            mbuf->ol_flags |=
+                (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
+            th->check = rte_ipv4_phdr_cksum(ip4_hdr(mbuf), mbuf->ol_flags);
+        }
     } else {
         if (mbuf_may_pull(mbuf, mbuf->pkt_len) != 0)
             return EDPVS_INVPKT;
-        tcp4_send_csum(ip4_hdr(mbuf), th);
+        if (af == AF_INET6) {
+            tcp6_send_csum((struct ipv6_hdr *)ip6_hdr(mbuf), th);
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_tcp.c]: 这个v6的转换是因为v6用了linux的包头格式，而tcp6_send_csum接收的是dpdk的v6格式

File: src/ipvs/ip_vs_proto_udp.c [hunk 17]
```diff
@@ -485,19 +602,26 @@ static int udp_fnat_out_handler(struct dp_vs_proto *proto,
...
     struct udp_hdr *uh;
+    int af = conn->af;
+    int iphdrlen = ((af == AF_INET6) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
 
     /* cannot use mbuf_header_pointer() */
-    if (unlikely(mbuf->data_len < ip4_hdrlen(mbuf) + sizeof(struct udp_hdr)))
+    if (unlikely(mbuf->data_len < iphdrlen + sizeof(struct udp_hdr)))
         return EDPVS_INVPKT;
-    uh = rte_pktmbuf_mtod_offset(mbuf, struct udp_hdr *, ip4_hdrlen(mbuf));
+    uh = rte_pktmbuf_mtod_offset(mbuf, struct udp_hdr *, iphdrlen);
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: 同上，update ip6_hdrlen

File: src/ipvs/ip_vs_proto_udp.c [hunk 18]
```diff
@@ -397,7 +496,7 @@ static int udp_insert_uoa(struct dp_vs_conn *conn, struct rte_mbuf *mbuf,
     }
 
     /* get udp header before any 'standalone_uoa' */
-    uh = rte_pktmbuf_mtod_offset(mbuf, struct udphdr *, ip4_hdrlen(mbuf));
+    uh = rte_pktmbuf_mtod_offset(mbuf, struct udphdr *, iphdrlen);
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: update ip6_hdrlen

File: src/ipvs/ip_vs_proto_tcp.c [hunk 19]
```diff
@@ -636,7 +636,7 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
         dev = conn->in_dev;
 
     if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
-        if (af == AF_INET6) {
+        if (AF_INET6 == af) {
             mbuf->l4_len = ntohs(ip6_hdr(mbuf)->ip6_plen);
             mbuf->l3_len = iphdrlen;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `l3_len` should  be NOT include exthdr lenght for IPv6 upper layer checksum.
See `https://tools.ietf.org/html/rfc2460#page-27` for detail.

File: src/ipvs/ip_vs_proto_tcp.c [hunk 20]
```diff
@@ -573,16 +591,17 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
     struct tcphdr *th;
     struct route_entry *rt = mbuf->userdata;
     struct netif_port *dev = NULL;
-    int ip4hlen = ip4_hdrlen(mbuf);
-    
-    if (mbuf_may_pull(mbuf, ip4hlen + sizeof(*th)) != 0)
+    int af = conn->af;
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `ip6_hdrlen()`  may return a negative.

File: src/ipvs/ip_vs_proto_tcp.c [hunk 21]
```diff
@@ -617,14 +636,28 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
         dev = conn->in_dev;
 
     if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
-        mbuf->l4_len = ntohs(ip4_hdr(mbuf)->total_length) - ip4hlen;
-        mbuf->l3_len = ip4hlen;
-        mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
-        th->check = rte_ipv4_phdr_cksum(ip4_hdr(mbuf), mbuf->ol_flags);
+        if (AF_INET6 == af) {
+            mbuf->l4_len = ntohs(ip6_hdr(mbuf)->ip6_plen);
+            mbuf->l3_len = iphdrlen;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: ` mbuf->l3_len` should consider exthdrs?

File: src/ipvs/ip_vs_proto_tcp.c [hunk 22]
```diff
@@ -636,16 +669,17 @@ static int tcp_fnat_out_handler(struct dp_vs_proto *proto,
     struct tcphdr *th;
     struct route_entry *rt = mbuf->userdata;
     struct netif_port *dev = NULL;
-    int ip4hlen = ip4_hdrlen(mbuf);
+    int af = conn->af;
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `ip6_hdrlen()` may return negative.

File: src/ipvs/ip_vs_proto_tcp.c [hunk 23]
```diff
@@ -671,14 +705,27 @@ static int tcp_fnat_out_handler(struct dp_vs_proto *proto,
         dev = conn->out_dev;
 
     if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
-        mbuf->l4_len = ntohs(ip4_hdr(mbuf)->total_length) - ip4hlen;
-        mbuf->l3_len = ip4hlen;
-        mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
-        th->check = rte_ipv4_phdr_cksum(ip4_hdr(mbuf), mbuf->ol_flags);
+        if (AF_INET6 == af) {
+            mbuf->l4_len = ntohs(ip6_hdr(mbuf)->ip6_plen);
+            mbuf->l3_len = iphdrlen;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `mbuf->l3_len` should consider exthdrs?

File: src/ipvs/ip_vs_proto_tcp.c [hunk 24]
```diff
@@ -688,18 +735,19 @@ static int tcp_snat_in_handler(struct dp_vs_proto *proto,
                                struct dp_vs_conn *conn, struct rte_mbuf *mbuf)
 {
     struct tcphdr *th;
-    int ip4hlen = ip4_hdrlen(mbuf);
     struct netif_port *dev = NULL;
     struct route_entry *rt = mbuf->userdata;
+    int af = conn->af;
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `ip6_hdrlen(mbuf)` may return a negative.

File: src/ipvs/ip_vs_proto_tcp.c [hunk 25]
```diff
@@ -711,14 +759,27 @@ static int tcp_snat_in_handler(struct dp_vs_proto *proto,
 
     /* leverage HW TX TCP csum offload if possible */
     if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
-        mbuf->l4_len = ntohs(ip4_hdr(mbuf)->total_length) - ip4hlen;
-        mbuf->l3_len = ip4hlen;
-        mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
-        th->check = rte_ipv4_phdr_cksum(ip4_hdr(mbuf), mbuf->ol_flags);
+        if (AF_INET6 == af) {
+            mbuf->l4_len = ntohs(ip6_hdr(mbuf)->ip6_plen);
+            mbuf->l3_len = iphdrlen;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `mbuf->l3_len` should consider exthdrs?

File: src/ipvs/ip_vs_proto_tcp.c [hunk 26]
```diff
@@ -728,18 +789,19 @@ static int tcp_snat_out_handler(struct dp_vs_proto *proto,
                                 struct dp_vs_conn *conn, struct rte_mbuf *mbuf)
 {
     struct tcphdr *th;
-    int ip4hlen = ip4_hdrlen(mbuf);
     struct netif_port *dev = NULL;
     struct route_entry *rt = mbuf->userdata;
+    int af = conn->af;
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `ip6_hdrlen(mbuf)` may return a negative

File: src/ipvs/ip_vs_proto_tcp.c [hunk 27]
```diff
@@ -751,14 +813,27 @@ static int tcp_snat_out_handler(struct dp_vs_proto *proto,
 
     /* leverage HW TX TCP csum offload if possible */
     if (likely(dev && (dev->flag & NETIF_PORT_FLAG_TX_TCP_CSUM_OFFLOAD))) {
-        mbuf->l4_len = ntohs(ip4_hdr(mbuf)->total_length) - ip4hlen;
-        mbuf->l3_len = ip4hlen;
-        mbuf->ol_flags |= (PKT_TX_TCP_CKSUM | PKT_TX_IP_CKSUM | PKT_TX_IPV4);
-        th->check = rte_ipv4_phdr_cksum(ip4_hdr(mbuf), mbuf->ol_flags);
+        if (AF_INET6 == af) {
+            mbuf->l4_len = ntohs(ip6_hdr(mbuf)->ip6_plen);
+            mbuf->l3_len = iphdrlen;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `mbuf->l3_len` should consider exthdrs?

File: src/ipvs/ip_vs_proto_tcp.c [hunk 28]
```diff
@@ -795,12 +870,14 @@ static int tcp_state_trans(struct dp_vs_proto *proto, struct dp_vs_conn *conn,
...
     struct dp_vs_dest *dest = conn->dest;
     unsigned conn_timeout = 0;
+    int af = conn->af;
 #ifdef CONFIG_DPVS_IPVS_DEBUG
     char dbuf[64], cbuf[64];
     const char *daddr, *caddr;
 #endif
 
-    th = mbuf_header_pointer(mbuf, ip4_hdrlen(mbuf), sizeof(_tcph), &_tcph);
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_tcp.c]: `ip6_hdrlen(mbuf)` may return a negative.

File: src/ipvs/ip_vs_proto_tcp.c (hunk 20 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_tcp.c]: 那是 ip6_skip_exthdr 要保证的

File: src/ipvs/ip_vs_proto_udp.c [hunk 29]
```diff
@@ -189,59 +200,100 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
         if (unlikely(!opp))
             goto no_room;
 
         memset(opp, 0, sizeof(*opp));
-        opp->version = 0x01;
-        opp->protocol = oiph->protocol;
-        opp->length = htons(sizeof(*opp) + sizeof(*uoa));
+        if (AF_INET6 == af) {
+            opp->version  = OPPHDR_IPV6;
+            opp->protocol = ((struct ip6_hdr *)oiph)->ip6_nxt;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: Is `oiph->ip6_nxt`  IPPROTO_UDP when exthdrs exist?

File: src/ipvs/ip_vs_proto_udp.c [hunk 30]
```diff
@@ -344,32 +421,51 @@ static int insert_opp_uoa(struct dp_vs_conn *conn, struct rte_mbuf *mbuf,
...
-    opph = (struct opphdr *)((void *)niph + (niph->ihl << 2));
+    opph = (struct opphdr *)((void *)niph + iphdrlen);
     memset(opph, 0, sizeof(*opph));
-    opph->version   = 0x1;
-    opph->protocol  = niph->protocol;
-    opph->length    = htons(sizeof(*opph) + IPOLEN_UOA);
+    if (AF_INET6 == af) {
+        /* version 2 for ipv6 address family */
+        opph->version  = OPPHDR_IPV6;
+        opph->protocol = ((struct ip6_hdr *)niph)->ip6_nxt;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: Is `niph->ip6_nxt` IPPROTO_UDP when exthdrs exits ?

File: src/ipvs/ip_vs_proto_udp.c [hunk 31]
```diff
@@ -441,26 +552,40 @@ static int udp_fnat_in_handler(struct dp_vs_proto *proto,
...
+    int af = conn->af;
+    int iphdrlen = 0;
+    uint8_t nxt_proto;
+
+    if (AF_INET6 == af) {
+        iph = ip6_hdr(mbuf);
+        iphdrlen = ip6_hdrlen(mbuf);
+        /* need found the last ip6_nxt of the ext header */
+        /* fix me */
+        nxt_proto = ((struct ip6_hdr *)iph)->ip6_nxt;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: use `ip6_skip_exthdr()` to get nxt_proto.

File: src/ipvs/ip_vs_proto_udp.c [hunk 32]
```diff
@@ -441,26 +552,40 @@ static int udp_fnat_in_handler(struct dp_vs_proto *proto,
...
 
-    if (iph->protocol == IPPROTO_UDP) {
-        uh = (void *)iph + ip4_hdrlen(mbuf);
-    } else if (iph->protocol == IPPROTO_OPT) {
-        opp = (void *)iph + ip4_hdrlen(mbuf);
-
-        uh = (void *)opp + ntohs(opp->length);
+    if (nxt_proto == IPPROTO_UDP) {
+        uh = (struct udp_hdr *)(iph + iphdrlen);
+    } else if (nxt_proto == IPPROTO_OPT) {
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: How does `ip6_skip_exthdr()` work with the private `IPPROTO_OPT` ? @AdamYYan

File: src/ipvs/ip_vs_proto_udp.c [hunk 33]
```diff
@@ -485,19 +616,26 @@ static int udp_fnat_out_handler(struct dp_vs_proto *proto,
                     struct rte_mbuf *mbuf)
 {
     struct udp_hdr *uh;
+    int af = conn->af;
+    int iphdrlen = ((AF_INET6 == af) ? ip6_hdrlen(mbuf): ip4_hdrlen(mbuf));
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: `ip6_hdrlen()` may return a negative.

File: src/ipvs/ip_vs_proto_udp.c [hunk 34]
```diff
@@ -519,18 +657,25 @@ static int udp_snat_in_handler(struct dp_vs_proto *proto,
...
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: `ip6_hdrlen()` may return a negative.

File: src/ipvs/ip_vs_proto_udp.c [hunk 35]
```diff
@@ -540,18 +685,25 @@ static int udp_snat_out_handler(struct dp_vs_proto *proto,
...
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_udp.c]: `ip6_hdrlen()` may return a negative.

File: src/ipvs/ip_vs_proto_udp.c [hunk 36]
```diff
@@ -189,59 +200,100 @@ static int send_standalone_uoa(const struct dp_vs_conn *conn,
...
+        ((struct iphdr *)iph)->daddr   = conn->daddr.in.s_addr;
+    }
 
     if (mode == UOA_M_IPO) {
-        iph->ihl    = (sizeof(struct iphdr) + IPOLEN_UOA) / 4;
-        iph->tot_len = htons(sizeof(*iph) + IPOLEN_UOA + sizeof(*uh));
-        iph->protocol = oiph->protocol; /* should always UDP */
-
-        uoa = (void *)rte_pktmbuf_append(mbuf, IPOLEN_UOA);
+        /* only ipv4 support and use this ip option mode */
```
[the remark this answers is not in the document]
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: ipv6 mode will not get here

File: src/ipvs/ip_vs_proto_udp.c (hunk 29 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: when in send_standalone_uoa, set to fixed IPPROTO_UDP

File: src/ipvs/ip_vs_proto_udp.c (hunk 33 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]: update `ip6_hdrlen()`, return 0 if negative

File: src/ipvs/ip_vs_proto_udp.c (hunk 35 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_udp.c]:
```c
/*
 * ip6_hdrlen: get ip6 header length, including extension header length
 */
int ip6_hdrlen(const struct rte_mbuf *mbuf) {
    struct ip6_hdr *ip6h = ip6_hdr(mbuf);
    uint8_t ip6nxt = ip6h->ip6_nxt;
    int ip6_hdrlen = ip6_skip_exthdr(mbuf, sizeof(struct ip6_hdr), &ip6nxt);

    /* ip6_skip_exthdr may return -1 */
    return (ip6_hdrlen >= 0) ? ip6_hdrlen : sizeof(struct ip6_hdr);
}
```

File: src/ipvs/ip_vs_proto_icmp.c (hunk 8 above)
Comment (xiangp126) [reply src/ipvs/ip_vs_proto_icmp.c]: icmp6 宏定义里只有这一种Reply

**Pull request merged**